zwrap/sample.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zwrap
+
+import (
+	"sync/atomic"
+	"time"
+
+	"github.com/uber-common/zap"
+)
+
+// Sample returns a sampling logger. Each tick, the sampler records the first N
+// messages and every Mth message thereafter. Sampling loggers are safe for
+// concurrent use.
+//
+// The children of sampling loggers (i.e., the results of calls to the With method)
+// inherit their parent's sampling logic but maintain independent counters.
+func Sample(zl zap.Logger, tick time.Duration, first, thereafter int) zap.Logger {
+	return &sampler{
+		Logger:     zl,
+		tick:       tick,
+		first:      uint64(first),
+		thereafter: uint64(thereafter),
```

Comment:
If I want to configure the Sampler to only print the first 100 messages, and then nothing else for that time period, what do you thik is cleaner? the user passing a large number for `thereafter` or making `thereafter == 0` mean that it should never log anything after `first`?